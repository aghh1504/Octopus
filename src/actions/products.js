import { normalize, schema } from "normalizr";

import productsSchema from "../schema/products";

import {
  SET_PRODUCTS_REQUEST,
  FETCH_PRODUCTS_REQUEST,
  SET_PRODUCTS_ERROR
} from "./types";

export const setProductsRequest = data => ({
  type: SET_PRODUCTS_REQUEST,
  payload: {
    data: normalize(data.data.products, new schema.Array(productsSchema))
  }
});

export const fetchProductsRequest = () => ({
  type: FETCH_PRODUCTS_REQUEST
});

export const setProductsErrors = data => ({
  type: SET_PRODUCTS_ERROR,
  payload: { data }
});
