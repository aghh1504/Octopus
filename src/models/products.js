import { Record } from "immutable";

const Product = Record(
  {
    id: null,
    title: "",
    image: {
      url: "",
      heigth: 0,
      width: 0
    },
    details: "",
    price: { amount: 0, currency: "" },
    quantity: 0,
    description: "",
    specifications: {
      brand: "",
      weight: "",
      dimensions: "",
      model_number: "",
      colour: "",
      type: "",
      socket: ""
    },
    min_rate: 0,
    max_rate: 0
  },
  "ProductRecord"
);

class ProductModel extends Product {}

export default ProductModel;

export function productFactory(props) {
  return new ProductModel(props);
}
