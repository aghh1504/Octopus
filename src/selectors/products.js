import { createSelector } from "../utils/reselect-memoize";

const getEntities = state => state.get("entities");

export const getProducts = createSelector(
  [getEntities],
  entities => entities.get("products")
);
