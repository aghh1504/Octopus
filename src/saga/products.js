import { put, takeLatest, call } from "redux-saga/effects";
import { get } from "./api";
import { BASE_URL } from "../config";

import { setProductsRequest, setProductsErrors } from "../actions/products";
import { FETCH_PRODUCTS_REQUEST } from "../actions/types";

export function* fetchProducts() {
  try {
    const products = yield call(get, `${BASE_URL}products`);
    yield put(setProductsRequest(products));
  } catch (e) {
    console.log(e);
    yield put(setProductsErrors(e.error));
  }
}

export function* watchProducts() {
  yield takeLatest(FETCH_PRODUCTS_REQUEST, fetchProducts);
}
