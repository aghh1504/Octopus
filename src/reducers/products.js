import { Map } from "immutable";

import { productFactory } from "../models/products";
import { SET_PRODUCTS_REQUEST } from "../actions/types";

const initState = Map();

export default function products(state = initState, action) {
  switch (action.type) {
    case SET_PRODUCTS_REQUEST: {
      const { products } = action.payload.data;
      const mappedProduct = products.reduce((accumulator, product) => {
        accumulator[product.id] = productFactory(product);
        return accumulator;
      }, {});
      return Map(mappedProduct);
    }

    default:
      return state;
  }
}
