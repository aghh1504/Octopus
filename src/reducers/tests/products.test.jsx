import products from "../products";

const productsModel = [
  {
    id: 1,
    image: {
      url: "",
      heigth: 510,
      width: 435
    },
    title: "Energy saving light bulb",
    details: "25W // Packet of 4",
    price: { amount: 12.99, currency: "Â£" },
    max_quantity: 10,
    description:
      "Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wiaint around warm start and flicker free features. Compatible with the Plumen Pendant. Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wiaint around warm start and flicker free features. Compatible with the Plumen Pendant.",
    specifications: {
      brand: "Philips",
      weight: "77g",
      dimensions: "12.6x6.2x6.2cm",
      model_number: "E27 ES",
      colour: "Cool daylight",
      type: "LED",
      socket: "B22/E27"
    },
    performance_first: [],
    performance_second: []
  }
];

const setProductsAction = {
  type: "SET_PRODUCTS_REQUEST",
  payload: {
    data: {
      entities: { products: productsModel }
    }
  }
};

describe("Products Reducer", () => {
  it("should store the products into state on the right event", () => {
    const result = products(undefined, setProductsAction);
    expect(result).toMatchSnapshot();
    result.forEach(product => {
      expect(product._name).toEqual("ProductRecord");
    });
  });
});
