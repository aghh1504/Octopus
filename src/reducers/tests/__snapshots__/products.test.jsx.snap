// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Reducer should store the products into state on the right event 1`] = `
Array [
  Immutable.ProductRecord {
    "id": 1,
    "title": "Energy saving light bulb",
    "image": Object {
      "heigth": 510,
      "url": "",
      "width": 435,
    },
    "details": "25W // Packet of 4",
    "price": Object {
      "amount": 12.99,
      "currency": "Â£",
    },
    "max_quantity": 10,
    "description": "Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wiaint around warm start and flicker free features. Compatible with the Plumen Pendant. Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wiaint around warm start and flicker free features. Compatible with the Plumen Pendant.",
    "specifications": Object {
      "brand": "Philips",
      "colour": "Cool daylight",
      "dimensions": "12.6x6.2x6.2cm",
      "model_number": "E27 ES",
      "socket": "B22/E27",
      "type": "LED",
      "weight": "77g",
    },
    "performance_first": Array [],
    "performance_second": Array [],
  },
]
`;
